container.go

```diff
@@ -196,16 +247,19 @@ func (c *Container) provideObject(o interface{}, otype reflect.Type) error {
 		otype = otype.Elem()
 		v = v.Elem()
 	}
+	c.insertObjectToGraph(v, otype)
+	return nil
+}
+
+func (c *Container) insertObjectToGraph(v reflect.Value, vtype reflect.Type) {
```

Comment:
The lock is applied in `Provide` and `Invoke` around insertObject func. I am refactoring this to move locking out and away from the container to simplify the API.